src/target/riscv/riscv-013.c

```diff
@@ -4103,7 +4141,8 @@ static int read_memory_progbuf_inner_one(struct target *target,
 	uint32_t command = access_register_command(target, GDB_REGNO_S1,
 			riscv_xlen(target), AC_ACCESS_REGISTER_WRITE |
 			AC_ACCESS_REGISTER_TRANSFER | AC_ACCESS_REGISTER_POSTEXEC);
-	if (execute_abstract_command(target, command) != ERROR_OK)
+	unsigned int abstractcs_err;
+	if (execute_abstract_command(target, command, &abstractcs_err) != ERROR_OK)
```

Comment:
> Would it be possible to make the third argument optional? If the user does not care about the exact cmderr value, they can pass NULL.

I'm reluctant to do this. I want to enforce programmer  to always think about error detection. We should not have situations where we don't care about errors.  We already had way too many commits to correct the situation (and more to come...)

Reply:
I fully agree that errors should always be taken into account. That however does not mean that the caller of `execute_abstract_command()` must always introspect the detailed cause of the error (cmderr) :-)

After thinking about it more, I believe it would be good to have 4 distinct return codes for execute_abstract_command():

- (1) `ERROR_OK` - Command succeeded. In this case the caller needs not look at cmderr (because cmderr is guaranteed to be 0).
- (2) `ERROR_CMD_FAILED` (or similar name) - Command was executed but failed due to cmderr!=0. The caller can look at cmderr to determine the actual failure reason, if it wishes, but it does not need to do so. All depends on the caller's situation, and how detailed information about the error it needs to have.
- (3) `ERROR_FAIL` - During the execution of the command, the DMI read or write operation failed. The target (or the connection to it) is seriously broken. The value in cmderr is irrelevant in this case.
- (4) `ERROR_TIMEOUT_REACHED` - Timeout occurred. The value in cmderr is not relevant in this case, either.

At the moment, cases (2) and (3) is merged into one ERROR_FAIL return code. It would be good to split it because those are two different outcomes. Please, let me know if you have similar or different opinion. But no need to make the changes in this merge request. It would just be good to agree on the strategy.

Based on the above, I still believe the third argument of execute_abstract_command() should be made optional. The caller should detect the overall status (1) - (4) through the return code. And only in case of (2), the cmderr can be introspected, but it is not mandatory to do so in all cases.